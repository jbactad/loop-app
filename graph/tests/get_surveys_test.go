package tests_test

import (
	"testing"

	"github.com/99designs/gqlgen/client"
	"github.com/gkampitakis/go-snaps/snaps"
	"github.com/jbactad/loop/application/ports/mocks"
	"github.com/stretchr/testify/assert"
)

func TestSurveys(t *testing.T) {
	type args struct {
		query string
		limit int
		page  int
	}
	tests := []struct {
		name    string
		args    args
		wantErr assert.ErrorAssertionFunc
	}{
		{
			name: "with valid query, it should return surveys without error",
			args: args{
				query: `query Surveys($limit: Int, $page: Int) {
  surveys(limit: $limit, page: $page) {
    id
    description
    name
    question
  }
}`,
				limit: 1,
				page:  0,
			},
		},
		{
			name: "with invalid query, it should return error",
			args: args{
				query: `query Surveys($limit: Int, $page: Int) {
					  surveys(limit: $limit, page: $page) {
						id
						description
						name
						question

						# This is invalid
						invalid
					  }
					}`,
				limit: 10,
				page:  0,
			},
			wantErr: assert.Error,
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			c := NewTestClient(t, mocks.NewTimeProvider(t), mocks.NewUUIDGenerator(t))

			var resp struct {
				Surveys []struct {
					ID          string
					Description string
					Name        string
					Question    string
					CreatedAt   string
					UpdatedAt   string
				}
			}

			err := c.Post(tt.args.query,
				&resp,
				client.Var("limit", tt.args.limit),
				client.Var("page", tt.args.page),
			)
			if tt.wantErr != nil {
				tt.wantErr(t, err)
				return
			}

			snaps.MatchJSON(t, resp)
		})
	}
}
