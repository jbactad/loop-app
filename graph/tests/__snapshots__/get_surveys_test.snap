
[TestSurveys/with_valid_query,_it_should_return_surveys_without_error - 1]
{
 "Surveys": [
  {
   "CreatedAt": "",
   "Description": "Survey 1 description",
   "ID": "8138abe3-06b1-5882-88dd-ba931b388f1d",
   "Name": "Survey 1",
   "Question": "What is your favorite color?",
   "UpdatedAt": ""
  }
 ]
}
---

[TestSurveys/with_invalid_query,_it_should_return_error - 1]
{
 "Surveys": null
}
---
